src/step/assemble_stiffness_matrix_springs.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Profiling_ScopedRegion.hpp>
+
+#include "src/elements/springs/springs.hpp"
+
+namespace openturbine {
+
+inline void AssembleStiffnessMatrixSprings(const Springs& springs) {
+    auto region = Kokkos::Profiling::ScopedRegion("Assemble Springs Stiffness");
+    Kokkos::parallel_for(
+        "AssembleSpringsStiffness", springs.num_elems,
+        KOKKOS_LAMBDA(const int i_elem) {
+            // Apply stiffness terms with appropriate signs for each node pair
+            static constexpr std::array<std::array<double, 2>, 2> node_signs = {{
+                {1., -1.},  // Signs for node 1
+                {-1., 1.}   // Signs for node 2
+            }};
+            for (auto i_node = 0U; i_node < 2U; ++i_node) {
+                for (auto j_node = 0U; j_node < 2U; ++j_node) {
+                    const double sign = node_signs[i_node][j_node];
+                    for (auto i_dof = 0U; i_dof < 3U; ++i_dof) {
+                        for (auto j_dof = 0U; j_dof < 3U; ++j_dof) {
+                            springs.stiffness_matrix_terms(i_elem, i_node, j_node, i_dof, j_dof) =
+                                sign * springs.a(i_elem, i_dof, j_dof);
+                        }
+                    }
+                }
+            }
```

Comment:
It may be best to unroll these.  Something like:
```
for (auto i_dof = 0U; i_dof < 3U; ++i_dof) {
    for (auto j_dof = 0U; j_dof < 3U; ++j_dof) {
        springs.stiffness_matrix_terms(i_elem, 0, 0, i_dof, j_dof) = springs.a(i_elem, i_dof, j_dof);
    }
}
for (auto i_dof = 0U; i_dof < 3U; ++i_dof) {
    for (auto j_dof = 0U; j_dof < 3U; ++j_dof) {
        springs.stiffness_matrix_terms(i_elem, 0, 1, i_dof, j_dof) = -springs.a(i_elem, i_dof, j_dof);
    }
}
for (auto i_dof = 0U; i_dof < 3U; ++i_dof) {
    for (auto j_dof = 0U; j_dof < 3U; ++j_dof) {
        springs.stiffness_matrix_terms(i_elem, 1, 0, i_dof, j_dof) = -springs.a(i_elem, i_dof, j_dof);
    }
}
for (auto i_dof = 0U; i_dof < 3U; ++i_dof) {
    for (auto j_dof = 0U; j_dof < 3U; ++j_dof) {
        springs.stiffness_matrix_terms(i_elem, 1, 1, i_dof, j_dof) = springs.a(i_elem, i_dof, j_dof);
    }
}
```

Reply:
Do you see any issues with the following?

```
for (auto i_dof = 0U; i_dof < 3U; ++i_dof) {
    for (auto j_dof = 0U; j_dof < 3U; ++j_dof) {
        springs.stiffness_matrix_terms(i_elem, 0, 0, i_dof, j_dof) =
            springs.a(i_elem, i_dof, j_dof);
        springs.stiffness_matrix_terms(i_elem, 0, 1, i_dof, j_dof) =
            -springs.a(i_elem, i_dof, j_dof);
        springs.stiffness_matrix_terms(i_elem, 1, 0, i_dof, j_dof) =
            -springs.a(i_elem, i_dof, j_dof);
        springs.stiffness_matrix_terms(i_elem, 1, 1, i_dof, j_dof) =
            springs.a(i_elem, i_dof, j_dof);
    }
}
```

Reply:
Yes. The order of traversal is an inefficient memory access pattern and will slow the kernel down a lot. You had a memory access pattern before, but I think we want to avoid a look up table or other encoding of the positive/negative signs. 

I was ok with the bad memory access pattern in the residual version because they were only 3 doubles away from each other, so everything will likely be loaded into cache the whole time anyways. I'm that case, I'm guessing the overhead of doing two loops would be higher.

Reply:
I refactored by unrolling the outer loop with separate for loops for each block. Let's make sure things look good there.